### PR TITLE
Slice 98

### PR DESCRIPTION
**Pull request by pawelkoper**

### CODE REVIEW

File: slice-core/src/main/java/com/cognifide/slice/core/internal/adapter/SliceAdapterFactory.java [hunk 1]
```diff
@@ -64,15 +65,15 @@ public SliceAdapterFactory(String injectorName, InjectorsRepository repository,
...
 		ContextProvider contextProvider = requestContextProvider.getContextProvider(injectorName);
 		if (contextProvider == null) {
-			ContextFactory factory = injector.getInstance(ContextFactory.class);
+			ContextFactory factory = injectorWithContext.getInstance(ContextFactory.class);
 			Context context = factory.getResourceResolverContext(resource.getResourceResolver());
 			contextProvider = new ConstantContextProvider(context);
 		}
-		injector.pushContextProvider(contextProvider);
-		return injector;
+        injectorWithContext.pushContextProvider(contextProvider);
```
Comment (trekawek) [inline slice-core/src/main/java/com/cognifide/slice/core/internal/adapter/SliceAdapterFactory.java]: Use tabs.

File: slice-core/src/main/java/com/cognifide/slice/core/internal/filter/ContextRequestFilter.java [hunk 2]
```diff
@@ -63,13 +66,12 @@
...
 // @formatter:on
-public class ContextRequestFilter implements Filter, RequestContextProvider {
-
-	@Reference
-	private InjectorsRepository injectorsRepo;
+public class ContextRequestFilter implements Filter, RequestContextProvider, InjectorLifecycleListener {
 
 	public static final int RANKING = -650;
 
+    private final List<String> injectorNames = new ArrayList<String>();
```
Comment (trekawek) [inline slice-core/src/main/java/com/cognifide/slice/core/internal/filter/ContextRequestFilter.java]: Use tabs.

File: slice-core/src/main/java/com/cognifide/slice/core/internal/filter/ContextRequestFilter.java [hunk 3]
```diff
@@ -124,4 +126,18 @@ public void init(final FilterConfig filterConfig) throws ServletException {
 	public void destroy() {
 	}
 
+    @Override
+    public void injectorCreated(Injector injector, InjectorConfig config) {
+        synchronized(this){
```
Comment (trekawek) [inline slice-core/src/main/java/com/cognifide/slice/core/internal/filter/ContextRequestFilter.java]: Use `CopyOnWriteArrayList` for the `injectorNames` and remove `synchronized()` statements.

File: slice-core/src/main/java/com/cognifide/slice/core/internal/adapter/SliceAdapterFactory.java (hunk 1 above)
Comment (pawelkoper) [inline slice-core/src/main/java/com/cognifide/slice/core/internal/adapter/SliceAdapterFactory.java]: fixed

File: slice-core/src/main/java/com/cognifide/slice/core/internal/filter/ContextRequestFilter.java (hunk 2 above)
Comment (pawelkoper) [inline slice-core/src/main/java/com/cognifide/slice/core/internal/filter/ContextRequestFilter.java]: fixed

File: slice-core/src/main/java/com/cognifide/slice/core/internal/filter/ContextRequestFilter.java (hunk 3 above)
Comment (pawelkoper) [inline slice-core/src/main/java/com/cognifide/slice/core/internal/filter/ContextRequestFilter.java]: fixed

**Comment by mmajchrzak** [comment]
Thx for this pull request. The solution will keep information about injectors in two places: InjectorRepository and ContextRequestFilter. This is not a good idea. I proposed another solution which also has some drawbacks, therefore we will be addressing issue with ContextRequestFilter in next Slice release. For now I'm closing this PR.

**Pull request closed without merging**